src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -404,7 +404,7 @@ - (void)_callNextReadyWorkItemWithContext:(ContextSnapshot const &)context
             case MTRBatchedFully:
                 MTR_LOG("MTRAsyncWorkQueue<%@> fully merged work item [%llu] into %llu",
                     context.description, nextWorkItem.uniqueID, workItem.uniqueID);
-                [_items removeObjectAtIndex:1];
+                [_items removeObjectAtIndex:firstNonRunningItemIndex];
```

Comment:
Nice catch! Could rename `firstNonRunningItemIndex` to `nextWorkItemIndex` to make it more obvious that that's the item we're removing here.